src/stats/pca.jl

```diff
@@ -0,0 +1,160 @@
+    center = zeros(Float64, indim)
+    lambda = zeros(Float64, outdim)
+    U      = zeros(Float64, indim, outdim)
+    CCIPCA(U, lambda, center, l, 0)
+end
+Base.length(o::CCIPCA) = outdim(o) # Number of eigen-vectors
+Base.getindex(o::CCIPCA, i) = o.U[:, i]
+Base.lastindex(o::CCIPCA) = outdim(o)
+Base.size(o::CCIPCA) = (indim(o), outdim(o))
+@inline outdim(o::CCIPCA) = length(o.lambda)
+@inline indim(o::CCIPCA) = length(o.center)
+function _fit!(o::CCIPCA, x::Vector{Float64})
+    @assert length(x) == indim(o)
+    n = o.n + 1
+    # update center with new observation:
+    #o.center = (o.n * o.center .+ x)/n
+    o.center += (x .- o.center)/n
+    # center the new observation, unless this is the first observation:
+    xi = (o.n > 0) ? (x .- o.center) : deepcopy(x)
+    # Now recalc eigen-values and -vectors given the new observation:
+    f = (1.0+o.l)/n
+    @inbounds for i in 1:outdim(o)
+        if i == n
+            o.lambda[i] = norm(xi)
+            o.U[:, i] = xi / o.lambda[i]
+            break
+        end
+        v = (1-f) * o.lambda[i] * o.U[:, i] + f * dot(o.U[:, i], xi) * xi
+        o.lambda[i] = norm(v)
+        o.U[:, i] = v/o.lambda[i]
+        xi = xi .- (dot(o.U[:, i], xi) * o.U[:, i])
+    end
+    o.n = n
+end
+"""
+    transform(o::CCIPCA, u::AbstractArray{Float64})
+
+Transform (i.e. project) the vector `u` into the PCA space 
+represented by `o`.
+"""
+function transform(o::CCIPCA, u::AbstractArray{Float64})
+    @assert indim(o) == length(u)
+    (u .- o.center)' * o.U
+end
+# We can also call the CCIPCA object itself to transform:
+(o::CCIPCA)(u::AbstractArray{Float64}) = transform(o, u)
+"""
+    reconstruct(o::CCIPCA, uproj::AbstractArray{Float64})
+
+Reconstruct the (projected) vector `uproj` back to the original
+space from which `o` has been fitted.
+"""
+function reconstruct(o::CCIPCA, uproj::AbstractArray{Float64})
+    @assert outdim(o) == length(uproj)
+    o.center .+ (o.U * uproj')
+end
+"""
+    fittransform!(o::CCIPCA, u::Vector{Float64})
+
+First `fit!` and then `transform` the vector `u` into the PCA
+space represented by `o`.
+"""
+function fittransform!(o::CCIPCA, u::Vector{Float64})
+    _fit!(o, u)
+    transform(o, u)
+end
+"""
+    eigenvalue(o::CCIPCA, i::Int)
+
+Get the `i`th eigenvalue of `o`.
+"""
+function eigenvalue(o::CCIPCA, i::Int)
+    @assert 1 <= i <= outdim(o) "This CCIPCA has $(outdim(o)) eigenvalues, cannot return eigenvalue $(i)!"
+    o.lambda[i]
+end
+"""
+    eigenvector(o::CCIPCA, i::Int)
+
+Get the `i`th eigenvector of `o`.
+"""
+function eigenvector(o::CCIPCA, i::Int)
+    @assert 1 <= i <= outdim(o) "This CCIPCA has $(outdim(o)) eigenvectors, cannot return eigenvector $(i)!"
+    o.U[:, i]
+end
+eigenvalues(o::CCIPCA) = Float64[eigenvalue(o, i) for i in 1:outdim(o)]
+"""
+    variation(o::CCIPCA)
+
+Get the variation (explained) in the direction of each eigenvector.
+Returns a vector of zeros if no vectors have yet been fitted.
+"""
+function variation(o::CCIPCA)
```

Comment:
Is this what MultivariateStats calls `principalvars`?  We should match the name if we can.

Reply:
Hmm, it is closely related but I don't think theirs is converted to a percent scale but is rather the eigenvalues. I'm not sure there is a standard term for this though, for example, the Wikipedia article on PCA says: "If component scores are standardized to unit variance, loadings must contain the data variance in them (and that is the magnitude of eigenvalues)." So I think the name in MultivariateStats is non-standard. Not sure how to address this, the one I propose here is practically useful since it can be used to adaptively decide how many dimensions to retain in a visualisation for example (no need to keep 3D if 2D explains 95% of the variance).

Reply:
Since the PCA in MultivariateStats saves also the residual variation not captured in the eigenvalues/vectors I'm not sure it makes sense to use the same name. But I'm ok with changing if you persist. :)

Reply:
Actually, this page indicates that the eigenvalues actually are the variances of the principal components:

https://stats.stackexchange.com/questions/346692/how-does-eigenvalues-measure-variance-along-the-principal-components-in-pca

so I'm ok calling it `principalvars` as in MultivariateStats and adding also `principalvar` like they do. I will keep the `eigenvalue` name also, though.

Reply:
Thus, the `variation` here is not the same as the `eigenvalue` or `principalvar`. However, to be consistent I will add an s to its name since it returns them all.

Reply:
Actually, `relativevariances` makes more sense and is more descriptive. A possible future extension would be to also save the residual variance which is not explained by any of the eigenvectors but since we do not calculate or save that now I suggest we go with this PR as-is.